File: py/nxs_writer.py
"""NXS Writer — direct-to-buffer .nxb emitter for Python.

Mirrors the Rust NxsWriter API:
    NxsSchema  — precompile keys once; share across NxsWriter instances.
    NxsWriter  — slot-based hot path; no per-field dict lookups during write.

Usage::

    from nxs_writer import NxsSchema, NxsWriter

    schema = NxsSchema(["id", "username", "score"])
    w = NxsWriter(schema)
    w.begin_object()
    w.write_i64(0, 42)
    w.write_str(1, "alice")
    w.write_f64(2, 9.5)
    w.end_object()
    data: bytes = w.finish()
"""
from __future__ import annotations

import struct
from typing import List

# ── Format constants ─────────────────────────────────────────────────────────

MAGIC_FILE    = 0x4E595842  # NYXB
MAGIC_OBJ     = 0x4E59584F  # NYXO
MAGIC_LIST    = 0x4E59584C  # NYXL
MAGIC_FOOTER  = 0x2153584E  # NXS!
VERSION       = 0x0101
FLAG_SCHEMA_EMBEDDED = 0x0002

_U8  = struct.Struct("<B")
_U16 = struct.Struct("<H")
_U32 = struct.Struct("<I")
_U64 = struct.Struct("<Q")
_I64 = struct.Struct("<q")
_F64 = struct.Struct("<d")

# ── MurmurHash3-64 ────────────────────────────────────────────────────────────

def _murmur3_64(data: bytes) -> int:
    MASK = 0xFFFFFFFFFFFFFFFF
    C1   = 0xFF51AFD7ED558CCD
    C2   = 0xC4CEB9FE1A85EC53
    h    = 0x93681D6255313A99
    length = len(data)
    i = 0
    while i < length:
        k = 0
        for b in range(8):
            if i + b < length:
                k |= data[i + b] << (b * 8)
        k = (k * C1) & MASK
        k ^= k >> 33
        h ^= k
        h = (h * C2) & MASK
        h ^= h >> 33
        i += 8
    h ^= length
    h ^= h >> 33
    h = (h * C1) & MASK
    h ^= h >> 33
    return h

# ── Schema ────────────────────────────────────────────────────────────────────

class NxsSchema:
    """Precompiled schema: maps key names to slot indices."""

    __slots__ = ("keys", "bitmask_bytes")

    def __init__(self, keys: List[str]) -> None:
        self.keys: List[str] = list(keys)
        # LEB128 bitmask byte count: (len + 6) // 7, minimum 1
        self.bitmask_bytes: int = max(1, (len(keys) + 6) // 7)

    def __len__(self) -> int:
        return len(self.keys)


# ── Frame (per open object) ───────────────────────────────────────────────────

class _Frame:
    __slots__ = ("start", "bitmask", "offset_table", "slot_offsets",
                 "last_slot", "needs_sort")

    def __init__(self, start: int, bitmask: bytearray) -> None:
        self.start = start
        self.bitmask = bitmask
        self.offset_table: List[int] = []   # relative offsets (from object start)
        self.slot_offsets: List[tuple] = [] # (slot, buf_off)
        self.last_slot = -1
        self.needs_sort = False


# ── Writer ────────────────────────────────────────────────────────────────────

class NxsWriter:
    """Slot-based .nxb emitter.

    Call ``begin_object()`` / typed write methods / ``end_object()`` for each
    record; call ``finish()`` to get the complete file bytes.
    """

    __slots__ = ("schema", "_buf", "_frames", "_record_offsets")

    def __init__(self, schema: NxsSchema) -> None:
        self.schema = schema
        self._buf = bytearray()
        self._frames: List[_Frame] = []
        self._record_offsets: List[int] = []

    # ── Object lifetime ──────────────────────────────────────────────────────

    def begin_object(self) -> None:
        """Open an object.  Must be balanced with end_object()."""
        schema = self.schema

        if not self._frames:
            self._record_offsets.append(len(self._buf))

        start = len(self._buf)

        # Build bitmask with LEB128 continuation bits pre-set
        bm = bytearray(schema.bitmask_bytes)
        for i in range(schema.bitmask_bytes - 1):
            bm[i] = 0x80  # continuation bit set; data bits clear

        self._frames.append(_Frame(start, bm))

        # Magic (4) + length placeholder (4)
        self._buf += _U32.pack(MAGIC_OBJ)
        self._buf += _U32.pack(0)  # back-patched in end_object

        # Reserve bitmask bytes
        self._buf += bytes(bm)

        # Reserve offset table: u16 per schema key (upper bound)
        self._buf += bytes(len(schema) * 2)

        # Align data area to 8 bytes from object start
        while (len(self._buf) - start) % 8 != 0:
            self._buf += b'\x00'

    def end_object(self) -> None:
        """Close the current object and back-patch the header."""
        if not self._frames:
            raise RuntimeError("end_object without begin_object")
        frame = self._frames.pop()

        total_len = len(self._buf) - frame.start

        # Back-patch Length at frame.start + 4
        _U32.pack_into(self._buf, frame.start + 4, total_len)

        # Back-patch bitmask at frame.start + 8
        bm_off = frame.start + 8
        for i, b in enumerate(frame.bitmask):
            self._buf[bm_off + i] = b

        # Back-patch offset table at frame.start + 8 + bitmask_bytes
        ot_start = bm_off + self.schema.bitmask_bytes
        present_count = len(frame.offset_table)

        if not frame.needs_sort:
            for i, rel in enumerate(frame.offset_table):
                _U16.pack_into(self._buf, ot_start + i * 2, rel)
        else:
            pairs = sorted(frame.slot_offsets, key=lambda x: x[0])
            for i, (_, buf_off) in enumerate(pairs):
                rel = buf_off - frame.start
                _U16.pack_into(self._buf, ot_start + i * 2, rel)

        # Zero unused offset-table slots
        used = present_count * 2
        reserved = len(self.schema) * 2
        for i in range(used, reserved):
            self._buf[ot_start + i] = 0

    def finish(self) -> bytes:
        """Assemble and return the complete .nxb file bytes."""
        if self._frames:
            raise RuntimeError("unclosed objects")

        schema_bytes = _build_schema(self.schema.keys)
        dict_hash    = _murmur3_64(schema_bytes)
        data_start_abs = 32 + len(schema_bytes)

        data_sector = bytes(self._buf)

        tail_ptr = data_start_abs + len(data_sector)
        tail    = _build_tail_index_records(data_start_abs, self._record_offsets, tail_ptr)

        out = bytearray()

        # Preamble (32 bytes)
        out += _U32.pack(MAGIC_FILE)
        out += _U16.pack(VERSION)
        out += _U16.pack(FLAG_SCHEMA_EMBEDDED)
        out += _U64.pack(dict_hash)
        out += _U64.pack(0)
        out += bytes(8)  # reserved

        out += schema_bytes
        out += data_sector
        out += tail

        return bytes(out)

    # ── Typed write methods ──────────────────────────────────────────────────

    def write_i64(self, slot: int, v: int) -> None:
        self._mark_slot(slot)
        self._buf += _I64.pack(v)

    def write_f64(self, slot: int, v: float) -> None:
        self._mark_slot(slot)
        self._buf += _F64.pack(v)

    def write_bool(self, slot: int, v: bool) -> None:
        self._mark_slot(slot)
        self._buf += bytes([0x01 if v else 0x00])
        self._buf += bytes(7)  # 7 padding bytes

    def write_time(self, slot: int, unix_ns: int) -> None:
        self._mark_slot(slot)
        self._buf += _I64.pack(unix_ns)

    def write_null(self, slot: int) -> None:
        self._mark_slot(slot)
        self._buf += bytes(8)

    def write_str(self, slot: int, v: str) -> None:
        self._mark_slot(slot)
        b = v.encode("utf-8")
        self._buf += _U32.pack(len(b))
        self._buf += b
        # pad to 8
        used = (4 + len(b)) % 8
        if used:
            self._buf += bytes(8 - used)

    def write_bytes(self, slot: int, data: bytes) -> None:
        self._mark_slot(slot)
        self._buf += _U32.pack(len(data))
        self._buf += data
        used = (4 + len(data)) % 8
        if used:
            self._buf += bytes(8 - used)

    def write_list_i64(self, slot: int, values: List[int]) -> None:
        self._mark_slot(slot)
        total = 16 + len(values) * 8
        self._buf += _U32.pack(MAGIC_LIST)
        self._buf += _U32.pack(total)
        self._buf += bytes([0x3D])  # SIGIL_INT '='
        self._buf += _U32.pack(len(values))
        self._buf += bytes(3)  # padding
        for v in values:
            self._buf += _I64.pack(v)

    def write_list_f64(self, slot: int, values: List[float]) -> None:
        self._mark_slot(slot)
        total = 16 + len(values) * 8
        self._buf += _U32.pack(MAGIC_LIST)
        self._buf += _U32.pack(total)
        self._buf += bytes([0x7E])  # SIGIL_FLOAT '~'
        self._buf += _U32.pack(len(values))
        self._buf += bytes(3)  # padding
        for v in values:
            self._buf += _F64.pack(v)

    # ── Convenience ──────────────────────────────────────────────────────────

    @staticmethod
    def from_records(keys: List[str], records: List[dict]) -> bytes:
        """Write records from plain dicts.  Returns complete .nxb bytes."""
        schema = NxsSchema(keys)
        w = NxsWriter(schema)
        for rec in records:
            w.begin_object()
            for i, key in enumerate(keys):
                if key not in rec:
                    continue
                val = rec[key]
                if val is None:
                    w.write_null(i)
                elif isinstance(val, bool):
                    w.write_bool(i, val)
                elif isinstance(val, int):
                    w.write_i64(i, val)
                elif isinstance(val, float):
                    w.write_f64(i, val)
                elif isinstance(val, str):
                    w.write_str(i, val)
                elif isinstance(val, (bytes, bytearray)):
                    w.write_bytes(i, bytes(val))
            w.end_object()
        return w.finish()

    # ── Private ──────────────────────────────────────────────────────────────

    def _mark_slot(self, slot: int) -> None:
        if not self._frames:
            raise RuntimeError("no active object")
        frame = self._frames[-1]

        # Set bitmask bit
        byte_idx = slot // 7
        bit_idx  = slot % 7
        frame.bitmask[byte_idx] |= (1 << bit_idx)

        # Record relative offset from object start
        rel = len(self._buf) - frame.start

        if slot < frame.last_slot:
            frame.needs_sort = True
        frame.last_slot = slot

        frame.offset_table.append(rel)
        frame.slot_offsets.append((slot, len(self._buf)))


# ── Module-level helpers ──────────────────────────────────────────────────────

def _build_schema(keys: List[str]) -> bytes:
    """Build the binary Schema Header bytes (padded to 8-byte boundary)."""
    encoded = [k.encode("utf-8") for k in keys]
    n = len(keys)

    # KeyCount(2) + TypeManifest(n) + null-terminated strings
    size = 2 + n + sum(len(e) + 1 for e in encoded)
    padded = size + ((-size) % 8)

    buf = bytearray(padded)
    p = 0
    _U16.pack_into(buf, p, n); p += 2
    for _ in keys:
        buf[p] = 0x22; p += 1  # '"' sigil
    for e in encoded:
        buf[p:p + len(e)] = e; p += len(e)
        buf[p] = 0x00;         p += 1
    # Remaining bytes already zero (bytearray init)
    return bytes(buf)


def _build_tail_index_records(data_start: int, record_offsets: List[int], tail_ptr: int) -> bytes:
    """Build the Tail-Index for all records."""
    n = len(record_offsets)
    # EntryCount(4) + N * [KeyID(2) + AbsOff(8)] + FooterTailPtr(8) + MagicFooter(4)
    buf = bytearray(4 + n * 10 + 12)
    p = 0

    _U32.pack_into(buf, p, n); p += 4

    for i, rel_off in enumerate(record_offsets):
        _U16.pack_into(buf, p, i); p += 2          # KeyID = record index
        _U64.pack_into(buf, p, data_start + rel_off); p += 8

    _U64.pack_into(buf, p, tail_ptr); p += 8

    # MagicFooter
    _U32.pack_into(buf, p, MAGIC_FOOTER)

    return bytes(buf)

File: js/nxs_writer.js
// NXS Writer — direct-to-buffer .nxb emitter for JavaScript.
//
// Mirrors the Rust NxsWriter API exactly:
//   NxsSchema  — precompile keys once; share across many NxsWriter instances.
//   NxsWriter  — slot-based hot path; no per-field Map lookups.
//
// Usage:
//   import { NxsSchema, NxsWriter } from "./nxs_writer.js";
//
//   const schema = new NxsSchema(["id", "username", "score"]);
//   const w = new NxsWriter(schema);
//   w.beginObject();
//   w.writeI64(0, 42n);        // slot index, BigInt for i64
//   w.writeStr(1, "alice");
//   w.writeF64(2, 9.5);
//   w.endObject();
//   const bytes = w.finish(); // Uint8Array

const MAGIC_FILE   = 0x4E595842; // NYXB
const MAGIC_OBJ    = 0x4E59584F; // NYXO
const MAGIC_LIST   = 0x4E59584C; // NYXL
const MAGIC_FOOTER = 0x2153584E; // NXS!
const VERSION      = 0x0101;
const FLAG_SCHEMA_EMBEDDED = 0x0002;

// ── MurmurHash3-64 ───────────────────────────────────────────────────────────

function murmur3_64(bytes, offset, length) {
  const C1 = 0xFF51AFD7ED558CCDn;
  const C2 = 0xC4CEB9FE1A85EC53n;
  const MASK = 0xFFFFFFFFFFFFFFFFn;
  let h = 0x93681D6255313A99n;
  const end = offset + length;
  for (let p = offset; p < end; p += 8) {
    let k = 0n;
    for (let i = 0; i < 8 && p + i < end; i++) {
      k |= BigInt(bytes[p + i]) << BigInt(i * 8);
    }
    k = (k * C1) & MASK;
    k ^= k >> 33n;
    h ^= k;
    h = (h * C2) & MASK;
    h ^= h >> 33n;
  }
  h ^= BigInt(length);
  h ^= h >> 33n;
  h = (h * C1) & MASK;
  h ^= h >> 33n;
  return h;
}

// ── Scratch buffer for f64 LE encoding ───────────────────────────────────────

const _scratchBuf = new ArrayBuffer(8);
const _scratchU8  = new Uint8Array(_scratchBuf);
const _scratchF64 = new Float64Array(_scratchBuf);

function encodeF64LE(v) {
  _scratchF64[0] = v;
  return _scratchU8.slice(0, 8);
}

// ── Schema ───────────────────────────────────────────────────────────────────

export class NxsSchema {
  /**
   * @param {string[]} keys
   */
  constructor(keys) {
    this.keys = keys;
    // LEB128 bitmask byte count for keys.length bits.
    // Same formula as Rust: integer (len + 6) / 7, minimum 1.
    // Math.floor mirrors Rust integer division exactly.
    this.bitmaskBytes = Math.max(1, Math.floor((keys.length + 6) / 7));
  }

  get length() { return this.keys.length; }
}

// ── Writer ───────────────────────────────────────────────────────────────────

// Sigil constants
const SIGIL_STR    = 0x22; // '"' — string / var-length
const SIGIL_I64    = 0x69; // 'i'
const SIGIL_F64    = 0x64; // 'd'
const SIGIL_BOOL   = 0x62; // 'b'
const SIGIL_NULL   = 0x6E; // 'n'
const SIGIL_BINARY = 0x42; // 'B'

export class NxsWriter {
  /**
   * @param {NxsSchema} schema
   */
  constructor(schema) {
    this.schema = schema;
    this._buf = new Uint8Array(4096);  // pre-allocated, grows by doubling
    this._view = new DataView(this._buf.buffer);
    this._len = 0;
    this._frames = [];       // stack of open object frames
    this._recordOffsets = []; // record start positions in data sector
    // Sigil per slot: default str/var-length; updated on each typed write
    this._slotSigils = new Uint8Array(schema.length).fill(SIGIL_STR);
  }

  // ── Buffer growth ────────────────────────────────────────────────────────

  _grow(need) {
    if (this._len + need <= this._buf.length) return;
    let newSize = this._buf.length * 2;
    while (this._len + need > newSize) newSize *= 2;
    const newBuf = new Uint8Array(newSize);
    newBuf.set(this._buf.subarray(0, this._len));
    this._buf = newBuf;
    this._view = new DataView(this._buf.buffer);
  }

  // ── Frame management ────────────────────────────────────────────────────

  beginObject() {
    const schema = this.schema;

    // Track top-level record offsets for the tail-index
    if (this._frames.length === 0) {
      this._recordOffsets.push(this._len);
    }

    const start = this._len;

    // Build initial bitmask with LEB128 continuation bits pre-set
    const bitmask = new Uint8Array(schema.bitmaskBytes);
    for (let i = 0; i < schema.bitmaskBytes - 1; i++) {
      bitmask[i] = 0x80;
    }
    // last byte has no continuation bit

    this._frames.push({
      start,
      bitmask,
      offsetTable: [],    // relative offsets from object start, in write order
      slotOffsets: [],    // [{slot, bufOff}] for sort path
      lastSlot: -1,
      needsSort: false,
    });

    // Write placeholder Magic (4) + Length (4) = 8 bytes
    this._writeU32(MAGIC_OBJ);
    this._writeU32(0); // length placeholder — back-patched in endObject

    // Reserve bitmask space
    this._writeBytes(bitmask);

    // Reserve offset-table space: one u16 per schema key (upper bound)
    const otReserve = new Uint8Array(schema.length * 2);
    this._writeBytes(otReserve);

    // Align data area to 8 from the object start
    while ((this._len - start) % 8 !== 0) {
      this._writeByte(0);
    }
  }

  endObject() {
    const frame = this._frames.pop();
    if (!frame) throw new Error("endObject without beginObject");

    // With a single contiguous buffer, back-patching is direct — no merge needed.
    const totalData = this._buf;

    const totalLen = this._len - frame.start;

    // Back-patch Length at frame.start + 4
    const lenOff = frame.start + 4;
    totalData[lenOff]     = totalLen & 0xFF;
    totalData[lenOff + 1] = (totalLen >>> 8) & 0xFF;
    totalData[lenOff + 2] = (totalLen >>> 16) & 0xFF;
    totalData[lenOff + 3] = (totalLen >>> 24) & 0xFF;

    // Back-patch bitmask at frame.start + 8
    const bmOff = frame.start + 8;
    for (let i = 0; i < frame.bitmask.length; i++) {
      totalData[bmOff + i] = frame.bitmask[i];
    }

    // Back-patch offset table starting at frame.start + 8 + bitmask.length
    const otStart = bmOff + frame.bitmask.length;
    const presentCount = frame.offsetTable.length;

    if (!frame.needsSort) {
      // Fast path: fields in slot order
      for (let i = 0; i < presentCount; i++) {
        const rel = frame.offsetTable[i];
        totalData[otStart + i * 2]     = rel & 0xFF;
        totalData[otStart + i * 2 + 1] = (rel >>> 8) & 0xFF;
      }
    } else {
      // Slow path: sort by slot, write offsets in slot order
      const pairs = frame.slotOffsets.slice().sort((a, b) => a.slot - b.slot);
      for (let i = 0; i < pairs.length; i++) {
        const rel = pairs[i].bufOff - frame.start;
        totalData[otStart + i * 2]     = rel & 0xFF;
        totalData[otStart + i * 2 + 1] = (rel >>> 8) & 0xFF;
      }
    }

    // Zero unused offset-table slots (deterministic output)
    const usedBytes = presentCount * 2;
    const reservedBytes = this.schema.length * 2;
    for (let i = usedBytes; i < reservedBytes; i++) {
      totalData[otStart + i] = 0;
    }
  }

  /**
   * Finish writing and return the complete .nxb file as a Uint8Array.
   */
  finish() {
    if (this._frames.length !== 0) throw new Error("unclosed objects");

    const schemaBytes = buildSchema(this.schema.keys, this._slotSigils);
    const dictHash    = murmur3_64(schemaBytes, 0, schemaBytes.length);

    // data_start_abs = 32 (preamble) + schemaBytes.length
    const dataStartAbs = 32 + schemaBytes.length;

    // Materialise data sector
    const dataSector = this._materializeAll();

    const tailPtr = dataStartAbs + dataSector.length;
    const tail    = buildTailIndexRecords(dataStartAbs, this._recordOffsets, tailPtr);

    const total = 32 + schemaBytes.length + dataSector.length + tail.length;
    const out = new Uint8Array(total);
    let p = 0;

    // Preamble (32 bytes)
    out[p++] = MAGIC_FILE & 0xFF;
    out[p++] = (MAGIC_FILE >>> 8) & 0xFF;
    out[p++] = (MAGIC_FILE >>> 16) & 0xFF;
    out[p++] = (MAGIC_FILE >>> 24) & 0xFF;

    out[p++] = VERSION & 0xFF;
    out[p++] = (VERSION >>> 8) & 0xFF;

    out[p++] = FLAG_SCHEMA_EMBEDDED & 0xFF;
    out[p++] = (FLAG_SCHEMA_EMBEDDED >>> 8) & 0xFF;

    // DictHash — 8 bytes little-endian BigInt (one-time, not hot path)
    for (let i = 0; i < 8; i++) {
      out[p++] = Number((dictHash >> BigInt(i * 8)) & 0xFFn);
    }

    // TailPtr=0 marks a streamable v1.1 file; the final tail pointer lives in the footer.
    {
      for (let i = 0; i < 8; i++) out[p++] = 0;
    }

    // Reserved 8 bytes
    p += 8;

    // Schema
    out.set(schemaBytes, p); p += schemaBytes.length;

    // Data sector
    out.set(dataSector, p); p += dataSector.length;

    // Tail-index
    out.set(tail, p);

    return out;
  }

  // ── Typed write methods ──────────────────────────────────────────────────

  writeI64(slot, v) {
    this._slotSigils[slot] = SIGIL_I64;
    this._markSlot(slot);
    let lo, hi;
    if (typeof v === "bigint") {
      const u = BigInt.asUintN(64, v);
      lo = Number(u & 0xFFFFFFFFn);
      hi = Number(u >> 32n);
    } else {
      // number path: handle signed 32-bit split correctly
      lo = v >>> 0;
      hi = Math.floor(v / 4294967296) >>> 0;
    }
    this._writeI64Parts(lo, hi);
  }

  writeF64(slot, v) {
    this._slotSigils[slot] = SIGIL_F64;
    this._markSlot(slot);
    this._writeBytes(encodeF64LE(v));
  }

  writeBool(slot, v) {
    this._slotSigils[slot] = SIGIL_BOOL;
    this._markSlot(slot);
    this._grow(8);
    this._buf[this._len] = v ? 0x01 : 0x00;
    this._buf.fill(0, this._len + 1, this._len + 8);
    this._len += 8;
  }

  writeTime(slot, unixNs) {
    this.writeI64(slot, unixNs);
  }

  writeNull(slot) {
    this._slotSigils[slot] = SIGIL_NULL;
    this._markSlot(slot);
    this._grow(8);
    this._buf.fill(0, this._len, this._len + 8);
    this._len += 8;
  }

  writeStr(slot, v) {
    this._slotSigils[slot] = SIGIL_STR;
    this._markSlot(slot);
    const bytes = encodeUtf8(v);
    const len = bytes.length;
    const used = (4 + len) % 8;
    const pad  = used === 0 ? 0 : (8 - used);
    this._grow(4 + len + pad);
    this._view.setUint32(this._len, len, true); this._len += 4;
    this._buf.set(bytes, this._len); this._len += len;
    if (pad > 0) { this._buf.fill(0, this._len, this._len + pad); this._len += pad; }
  }

  writeBytes(slot, data) {
    this._slotSigils[slot] = SIGIL_BINARY;
    this._markSlot(slot);
    const len = data.length;
    const used = (4 + len) % 8;
    const pad  = used === 0 ? 0 : (8 - used);
    this._grow(4 + len + pad);
    this._view.setUint32(this._len, len, true); this._len += 4;
    this._buf.set(data, this._len); this._len += len;
    if (pad > 0) { this._buf.fill(0, this._len, this._len + pad); this._len += pad; }
  }

  writeListI64(slot, values) {
    this._markSlot(slot); // list is var-length — keep SIGIL_STR default
    const total = 16 + values.length * 8;
    this._writeU32(MAGIC_LIST);
    this._writeU32(total);
    this._writeByte(0x3D); // SIGIL_INT '='
    this._writeU32(values.length);
    this._writeByte(0); this._writeByte(0); this._writeByte(0); // padding
    for (const v of values) {
      let lo, hi;
      if (typeof v === "bigint") {
        const u = BigInt.asUintN(64, v);
        lo = Number(u & 0xFFFFFFFFn);
        hi = Number(u >> 32n);
      } else {
        lo = v >>> 0;
        hi = Math.floor(v / 4294967296) >>> 0;
      }
      this._writeI64Parts(lo, hi);
    }
  }

  writeListF64(slot, values) {
    this._markSlot(slot);
    const total = 16 + values.length * 8;
    this._writeU32(MAGIC_LIST);
    this._writeU32(total);
    this._writeByte(0x7E); // SIGIL_FLOAT '~'
    this._writeU32(values.length);
    this._writeByte(0); this._writeByte(0); this._writeByte(0);
    for (const v of values) {
      this._writeBytes(encodeF64LE(v));
    }
  }

  // ── Convenience: write from record objects ───────────────────────────────

  /**
   * Write multiple records from plain JS objects.
   * @param {string[]} keys
   * @param {object[]} records
   * @returns {Uint8Array}
   */
  static fromRecords(keys, records) {
    const schema = new NxsSchema(keys);
    const w = new NxsWriter(schema);
    for (const rec of records) {
      w.beginObject();
      for (let i = 0; i < keys.length; i++) {
        const key = keys[i];
        if (!(key in rec)) continue;
        const val = rec[key];
        if (val === null || val === undefined) {
          w.writeNull(i);
        } else if (typeof val === "bigint") {
          w.writeI64(i, val);
        } else if (typeof val === "number") {
          if (Number.isSafeInteger(val)) {
            w.writeI64(i, val);
          } else {
            w.writeF64(i, val);
          }
        } else if (typeof val === "boolean") {
          w.writeBool(i, val);
        } else if (typeof val === "string") {
          w.writeStr(i, val);
        }
      }
      w.endObject();
    }
    return w.finish();
  }

  // ── Reset for WAL reuse ──────────────────────────────────────────────────

  reset() {
    this._len = 0;
    this._frames = [];
    this._recordOffsets = [];
    this._slotSigils.fill(SIGIL_STR);
  }

  // ── Private helpers ──────────────────────────────────────────────────────

  _markSlot(slot) {
    const frame = this._frames[this._frames.length - 1];
    if (!frame) throw new Error("no active object");

    // Set bitmask bit (slot index → byte/bit in LEB128 encoding)
    const byteIdx = Math.floor(slot / 7);
    const bitIdx  = slot % 7;
    frame.bitmask[byteIdx] |= (1 << bitIdx);

    // Record relative offset from object start
    const rel = this._len - frame.start;

    if (slot < frame.lastSlot) {
      frame.needsSort = true;
    }
    frame.lastSlot = slot;

    frame.offsetTable.push(rel);
    frame.slotOffsets.push({ slot, bufOff: this._len });
  }

  _writeByte(b) {
    this._grow(1);
    this._buf[this._len++] = b & 0xFF;
  }

  _writeU32(v) {
    this._grow(4);
    this._view.setUint32(this._len, v, true);
    this._len += 4;
  }

  _writeI64Parts(lo, hi) {
    this._grow(8);
    this._view.setUint32(this._len, lo, true);
    this._view.setUint32(this._len + 4, hi, true);
    this._len += 8;
  }

  _writeBytes(bytes) {
    if (bytes.length === 0) return;
    this._grow(bytes.length);
    this._buf.set(bytes, this._len);
    this._len += bytes.length;
  }

  /**
   * Return a view of the written bytes (no merge needed — single contiguous buffer).
   */
  _materialize() {
    return this._buf.subarray(0, this._len);
  }

  /**
   * Materialise and return the final data-sector bytes.
   */
  _materializeAll() {
    return this._materialize();
  }
}

// ── Helpers ──────────────────────────────────────────────────────────────────

const _utf8Encoder = new TextEncoder();

function encodeUtf8(s) {
  return _utf8Encoder.encode(s);
}

function buildSchema(keys, sigils) {
  const keyCount = keys.length;
  // Encode all keys as UTF-8 bytes
  const encoded = keys.map(k => _utf8Encoder.encode(k));

  // Size: u16 + keyCount bytes (type manifest) + null-terminated strings
  let size = 2 + keyCount; // KeyCount u16 + TypeManifest
  for (const e of encoded) size += e.length + 1; // string + null terminator
  // Pad to 8-byte boundary
  const padded = size + ((8 - (size % 8)) % 8);

  const buf = new Uint8Array(padded);
  let p = 0;

  // KeyCount u16 LE
  buf[p++] = keyCount & 0xFF;
  buf[p++] = (keyCount >>> 8) & 0xFF;

  // TypeManifest: emit per-slot sigil
  for (let i = 0; i < keyCount; i++) buf[p++] = sigils[i];

  // StringPool
  for (const e of encoded) {
    buf.set(e, p); p += e.length;
    buf[p++] = 0x00; // null terminator
  }
  // Remaining bytes are already zero (padding)

  return buf;
}

function buildTailIndexRecords(dataStartAbs, recordOffsets, tailPtr) {
  const n = recordOffsets.length;
  // EntryCount(4) + N * [KeyID(2) + AbsOff(8)] + FooterTailPtr(8) + MagicFooter(4)
  const buf = new Uint8Array(4 + n * 10 + 12);
  let p = 0;

  // EntryCount
  buf[p++] = n & 0xFF;
  buf[p++] = (n >>> 8) & 0xFF;
  buf[p++] = (n >>> 16) & 0xFF;
  buf[p++] = (n >>> 24) & 0xFF;

  for (let i = 0; i < n; i++) {
    // KeyID = record index (u16)
    buf[p++] = i & 0xFF;
    buf[p++] = (i >>> 8) & 0xFF;
    // AbsoluteOffset (u64) = dataStartAbs + relOff
    const abs = dataStartAbs + recordOffsets[i];
    // Write as 8-byte LE number (safe up to 2^53)
    const lo = abs >>> 0;
    const hi = Math.floor(abs / 0x100000000);
    buf[p++] = lo & 0xFF;
    buf[p++] = (lo >>> 8) & 0xFF;
    buf[p++] = (lo >>> 16) & 0xFF;
    buf[p++] = (lo >>> 24) & 0xFF;
    buf[p++] = hi & 0xFF;
    buf[p++] = (hi >>> 8) & 0xFF;
    buf[p++] = (hi >>> 16) & 0xFF;
    buf[p++] = (hi >>> 24) & 0xFF;
  }

  {
    const lo = tailPtr >>> 0;
    const hi = Math.floor(tailPtr / 0x100000000);
    buf[p++] = lo & 0xFF;
    buf[p++] = (lo >>> 8) & 0xFF;
    buf[p++] = (lo >>> 16) & 0xFF;
    buf[p++] = (lo >>> 24) & 0xFF;
    buf[p++] = hi & 0xFF;
    buf[p++] = (hi >>> 8) & 0xFF;
    buf[p++] = (hi >>> 16) & 0xFF;
    buf[p++] = (hi >>> 24) & 0xFF;
  }

  // MagicFooter
  buf[p++] = MAGIC_FOOTER & 0xFF;
  buf[p++] = (MAGIC_FOOTER >>> 8) & 0xFF;
  buf[p++] = (MAGIC_FOOTER >>> 16) & 0xFF;
  buf[p++] = (MAGIC_FOOTER >>> 24) & 0xFF;

  return buf;
}
